doc/architectural/memory-primitives.md

```diff
@@ -91,6 +91,69 @@ point into another memory object (such as could happen when running on a real
 machine). To verify that pointers stay within the bounds of their pointees, the
 CBMC option `--pointer-overflow-check` can be used.
 
+### Malloc modelling
+
+CBMC ships a model of `malloc` that seeks to emulate the behaviour of the C
+standard library. In an actual execution, memory allocation may fail for a
+number of reasons and `malloc` would return a NULL pointer. CBMC's model,
+however, will always succeed to allocate an object and never return the `NULL`
+pointer, unless configured as described below. This extends to the case of
```

Comment:
I would give a bit of rationale here -- this otherwise sounds like a bug. This is likely owed to the fact that code written for Linux benefits from the optimistic allocation of memory pages, and thus, programmers often omit the check for NULL. Plus, I'd say right away that this is the default and that the model can be configured, and that the appropriate setting depends on whether the code has NULL checks or not.

Reply:
I have attempted to rewrite this based on your feedback.